README.md
wordcloud
=========

Word Cloud activity for Sugar

Paste text into the entry and hit start to generate a word cloud that is saved to the Sugar Journal.

Based on the pytagcloud package.
![WordCloud](https://github.com/sugarlabs/wordcloud/blob/master/WordCloud.png)


NEWS
NEWS
7
* Fix labelling error (Sourabha G)
* Feature: Accepts uppercase characters (Sourabha G)
* Remove unnecessary stop button accelerator (James Cameron)
* Add number to image name when writing to datastore (Ibiam Chihurumnaya)
* Fix broken image link (Rahul Vaish)

6
* Port to Python3 (Kiy4h, Jui Pradhan)
* Fix IndexError: pop from empty list (Jui Pradhan)
* Change the undo counter (Jui Pradhan)
* flake8 fixes (Jui Pradhan)
* Port from GObject to GLib (Jui Pradhan)
* Only copy clipboard if there is a currently selected text (Ibiam Chihurumnaya)
* Modify shebang (Jui Pradhan)
* Fix TypeError: decoding str is not supported (Jui Pradhan)
* Fit words on display of all ratios (Jui Pradhan)
* Add get_display_rectangle() to return Screen width and height (James Cameron)
* Fix Deprecation Warning (Jui Pradhan)
* Regenerate POT file (Jui Pradhan)

5
* Fine tune repeated word mechanism (Walter Bender)
* Add a border (Walter Bender)
* More fine-tuning of padding; repeating (Walter Bender)
* Add XO color scheme (Walter Bender)
* Pep8 cleanup (Walter Bender)
* Make XO colors the default; New icon (Walter Bender)
* Add one-education colors (Walter Bender)
* Make color palette on a 3x3 grid (Walter Bender)
* Generate color bitmaps dynamically (Walter Bender)
* Run wordcloud stand-alone to end-run memory leak (Walter Bender)
* More robus stopword checks (Walter Bender)
* Add Alerts (Walter Bender)
* Highlight selected layout (Walter Bender)
* Add highlights to color palette (Walter Bender)
* Smaller canvas (Walter Bender)
* New artwork from dan (Walter Bender)
* Cleaned up artwork (Walter Bender)
* Clean up padding (Walter Bender) 
* Fine-tune colors (Walter Bender) 
* New strings (Walter Bender)
* Save text input between sessions (Walter Bender)
* Highlight font selection (Walter Bender)
* Add category (Walter Bender)
* Add screenshots and fix metadata (Aryan Sharma)
* Tidy up logging, fix indents (Zakiyah)
* Fix undo and Redo (Jui Pradhan)
* Regenerate POT file (Ibiam Chihurumnaya)

4
ENHANCEMENTS:
* Added layout options: horizontal, vertical, mixed, fortyfive, random
* Added color options
* Added repeated words toggle
* New icons
* Use white background for image

3
BUG FIX:
* Fixed font menu palette

2

ENHANCEMENTS:
* New Spanish translations (thanks to Irenarco)
* Load image on start
* Font selector

BUG FIX:
* Unicode fonts

1

New project


foo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from pytagcloud import create_tag_image, make_tags
from pytagcloud.lang.counter import get_tag_counts
 
TEXT = '''
Why is Free/Libre Software Important to Education?


The Expanding Role of Computing in Education

In a 2009-10 study in Montevideo, primary-school students used computers not just for playing games, but also for reading, writing, chatting, making music and videos, drawing, searching for information, and programming. They played -- they are children after all -- but they also used computation as a tool for communication -- “a thing with which to be social and expressive ” -- and problem-solving -- “a thing to think with”. Skills that they will be able to draw upon not just in school but in life.

How did this happen and how do we further promote and sustain it?


The True Cost of Proprietary Software

Change is the one thing that is consistent in computing. Most of the knowledge of IT taught in primary and middle schools some years ago is completely deprecated due to rapid changes in software and devices. This necessitates that software, knowledge and techniques should be updated frequently. 

A frequently cited reason for using Free/Libre Software is that is it free: upgrades are available at no cost. It is true that Free Software allows schools to free themselves of the costs associated with proprietary software and reduce the dependence of software vendor. However, cost savings is a Red Herring: there are maintenance costs associated with all software, regardless of whether or not one has to pay a licensing fee. (One upside to Free/Libre Software is that the investment you make in maintenance can be used as a source of nurturing local talent.)

The hidden cost of proprietary software is more insidious: it takes away from the user the opportunity to achieve autonomy, mastery, and a sense of purpose. In the context of school, it channels education into consumption, passivity, and acceptance. Free/Libre Software, on the other hand, focuses the student on creativity, activity, and intellectual risk taking. The student (and teacher) has the license to learn and, when properly staged, provides the means to leverage the opportunities afforded by the license.


A Culture of Autonomy and Responsibility

The “free” in Free/Libre Software refers to freedom. Users have the license to use, share or modify it according to their needs. The reality is that even for small software companies, an individual school or classroom is too small of a market to consider important. With Free/Libre Software, you don’t have to depend on third parties to do things for you. You are empowered to shape the tools to meet your own needs. It is only way to be freed from external dependencies. (Because your can share Free/Libre Software, you need not act alone: you can work within a community; communities arise where ever Free/Libre Software is used and are often global in reach). With autonomy comes responsibility: thus the culture of Free/Libre Software is a culture of individualism coupled with personal and communal responsibility. Not bad traits to instill in our children.


The Means of Mastery

With Free/Libre Software there is an increased opportunity to provide direct feedback between users and developers since there is no reason to build a “firewall” between them. This means that learners and educators can collaborate together with software developers in the design of educational software and create software which fits better in classrooms. A common forum for this interaction is “chat”. Many Free/Libre Software projects use open chat-rooms as a place for developers to discuss their work. Anyone is welcome to “listen” as domain experts openly discuss the problems with which they are wrestling. 

The interaction between learners and domain experts makes possible the transition from a learner to a software developer and become a contributor of a Free/Libre Software project. The Free/Libre Software community is a meritocracy: all that matters is what you do, not who you are. So there is room for personal growth. There is an open door to mastery within Free/Libre Software that is usually closed with proprietary software. Free/Libre software gives everyone the opportunity to be both a learner and a teacher.


A Sense of Purpose

Nothing motivates learning more than a sense of purpose. Neither carrots or sticks come close. Free/Libre Software gives the learner the opportunity to “scratch an itch” -- to pursue a problem of passion and purpose. It is while engaging in authentic problem-solving where one acquires the problem-solving skills that transfer across domains and are applicable beyond the walls of the classroom.

Within the Free/Libre Software community, we do not passively accept the status quo as inevitable, but actively embark upon making things better. Instilling that habit of mind in the next generation is perhaps the most valuable contribution Free/Libre Software makes to Education.


A Call to Action

The cultivation of Free/Libre Software in education doesn’t happen on its own. The proprietary software industry will fight to maintain a model of schools and learners as consumers. We need to protect our freedom to create. It is our responsibility. It won’t be done for us. If we don’t invest in the next generation of leadership, we will be overwhelmed.

The Free/Libre Software movement needs to foster growth and leadership among the next generation. We need to provide both the license and the means to exercise this license.

The next generation might have reasons to spread the Free/Libre Software culture, because they will be educated in that way. The most important factors are the deployment of Free/Libre Software in classrooms and the existence of an open community that facilitates feedback with schools.
'''
 
tags = make_tags(get_tag_counts(TEXT), maxsize=150)
 
create_tag_image(tags, 'cloud_large.png', size=(900, 600))


wordcloud.py
#! /usr/bin/env python3
# wordclound.py
"""
    Copyright (C) 2014 Walter Bender

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

"""

from pytagcloud import create_tag_image, make_tags, LAYOUT_MIX
from pytagcloud.lang.counter import get_tag_counts

import sys
import os
import gi

gi.require_version('Gdk', '3.0')

from gi.repository import Gdk

from io import StringIO

import json
json.dumps
from json import load as jload
from json import dump as jdump


def json_load(text):
    """ Load JSON data using what ever resources are available. """
    # strip out leading and trailing whitespace, nulls, and newlines
    io = StringIO(text)
    try:
        listdata = jload(io)
    except ValueError:
        # assume that text is ascii list
        listdata = text.split()
        for i, value in enumerate(listdata):
            listdata[i] = int(value)
    return listdata


def json_dump(data):
    """ Save data using available JSON tools. """
    _io = StringIO()
    jdump(data, _io)
    return _io.getvalue()


class WordCloud():

    def __init__(self):
        try:
            fd = open('/tmp/cloud_data.txt', 'r')
            data = fd.read()
            fd.close()
            cloud_dict = json_load(data)
            self._repeat_tags = cloud_dict['repeat']
            self._layout = cloud_dict['layout']
            self._font_name = cloud_dict['font']
            self._color_scheme = cloud_dict['colors']
        except:
            self._repeat_tags = False
            self._layout = LAYOUT_MIX
            self._font_name = None
            self._color_scheme = ((241, 143, 0), (128, 186, 39),
                                  (13, 147, 210), (231, 30, 108),
                                  (135, 135, 135))
        try:
            fd = open('/tmp/cloud_text.txt', 'r')
            self._text = fd.read()
            fd.close()
        except:
            self._text = 'Could not read cloud_text'

    def run(self):
        self._create_image(self._text)

    def get_display_rectangle(self):
        display = Gdk.Display.get_default()
        monitor = display.get_primary_monitor()
        geometry = monitor.get_geometry()
        scale_factor = monitor.get_scale_factor()
        width = scale_factor * geometry.width
        height = scale_factor * geometry.height
        return (width, height)

    def _create_image(self, text):
        tag_counts = get_tag_counts(text)
        if tag_counts is None:
            sys.exit(-1)

        if self._repeat_tags:
            expanded_tag_counts = []
            for tag in tag_counts:
                expanded_tag_counts.append((tag[0], 5))
            for tag in tag_counts:
                expanded_tag_counts.append((tag[0], 2))
            for tag in tag_counts:
                expanded_tag_counts.append((tag[0], 1))
            tag_counts = expanded_tag_counts

        tags = make_tags(tag_counts, maxsize=150, colors=self._color_scheme)
        path = os.path.join('/tmp/cloud_large.png')

        width, height = self.get_display_rectangle()

        if height < width:
            ratio = width / height
            width = int(height * ratio)
        else:
            ratio = height / width
            height = int(width * ratio)

        if self._font_name is not None:
            create_tag_image(tags, path, layout=self._layout,
                             size=(width, height),
                             fontname=self._font_name)
        else:
            create_tag_image(tags, path, layout=self._layout,
                             size=(width, height))
        return 0


if __name__ == "__main__":
    game = WordCloud()
    game.run()
    sys.exit(0)


pytagcloud/colors.py
# -*- coding: utf-8 -*-
COLOR_SCHEMES = {
    'oldschool': ((59, 76, 76), (125, 140, 116), (217, 175, 95), (127, 92, 70), 
                  (51, 36, 35)), 
    'citrus': ((34, 51, 49), (70, 102, 66), (153, 142, 61), (229, 156, 44), 
               (255, 116, 37)), 
    'goldfish': ((229, 106, 0), (204, 199, 148), (153, 145, 124), (88, 89, 86), 
                 (48, 49, 51)), 
    'audacity': ((181, 40, 66), (255, 195, 82), (255, 138, 57), (239, 93, 74), 
                 (90, 0, 82)), 
    'oneeducation': ((241, 143, 0), (128, 186, 39), (13, 147, 210), 
                     (231, 30, 108), (135, 135, 135)), 
    'munsell': ((242, 0, 60), (239, 204, 0), (0, 168, 119), (0, 147, 175),
                (159, 0, 197)),
    'munsell2': ((242, 136, 0), (177, 215, 0), (0, 165, 156), (0, 110, 191),
                 (208, 0, 129)),
    'blue': ((222, 83, 255), (143, 82, 232), (103, 103, 255), (82, 141, 232),
             (91, 219, 255)),
    'water': ((238, 57, 92), (242, 164, 38), (233, 241, 224), (85, 216, 217),
               (66, 177, 190)),
    'boom': ((152, 191, 34), (62, 76, 83), (228, 67, 46), (229, 147, 37),
             (48, 219, 231)),
    'crafty': ((117, 18, 21), (57, 141, 157), (112, 193, 208), (228, 254, 180),
               (156, 34, 88)),
    'expo': ((68, 87, 113), (55, 164, 167), (251, 169, 70), (249, 46, 71),
             (204, 36, 88)),
    'watermelon': ((85, 0, 3), (251, 84, 75), (70, 95, 11), (228, 254, 180),
                   (35, 120, 37)),
}



pytagcloud/lang/counter.py
# -*- coding: utf-8 -*-
import re
from pytagcloud.lang.stopwords import StopWords
from operator import itemgetter
import logging

def get_tag_counts(text):
    """
    Search tags in a given text. The language detection is based on stop lists.
    This implementation is inspired by https://github.com/jdf/cue.language. Thanks Jonathan Feinberg.
    """
    text = str(text)
    words = re.findall(r'\w+', text, re.UNICODE)

    # Workaround because the regex mangled the unicode
    # _PUNCTUATION = u'.,<>;:/?\\\"\'-_+=!@\#$%^&*()'
    # words = text.split()
    # words = map(lambda x:x.lower(), words)
    # words = map(lambda x:unicode(x, 'UTF-8'), words)
    # words = map(lambda x: x.translate(dict.fromkeys(map(ord, _PUNCTUATION))),
    #             words)

    s = StopWords()
    s.load_language(s.guess(words))
    
    counted = {}
    
    for word in words:
        if not s.is_stop_word(word) and len(word) >= 1:
            if word in counted:
                counted[word] += 1
            else: 
                counted[word] = 1
      
    if len(list(counted.keys())) == 0:
        return None
    else:
        return sorted(iter(counted.items()), key=itemgetter(1), reverse=True)


pytagcloud/lang/stopwords.py
# -*- coding: utf-8 -*-
import os

# ACTIVE_LISTS = ('german', 'french', 'italian', 'english', 'spanish')
ACTIVE_LISTS = ('english', 'spanish')

class StopWords(object):
    
    def __init__(self):
        
        self.stop_words_lists = {}
        self.language = None
        
        stop_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'stop')
        
        for root, dirs, files in os.walk(stop_dir):
            for file in files:
                if not file in ACTIVE_LISTS:
                    continue
                stop_file = open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'stop/', file), 'r')                
                self.stop_words_lists[file] = []                
                for stop_word in stop_file:
                    self.stop_words_lists[file].append(stop_word.strip().lower())
                stop_file.close()

    def load_language(self, language):
        self.language = language
                    
    def is_stop_word(self, word):
        if not self.language:
            raise LookupError("No language loaded")
        return word in self.stop_words_lists[self.language]
    
    def guess(self, words):
        currentWinner = ACTIVE_LISTS[0];
        currentMax = 0;
        
        for language, stop_word_list in list(self.stop_words_lists.items()):
            count = 0
            for word in words:
                if word.encode('ascii', 'ignore') in stop_word_list:
                    count += 1
                    
            if count > currentMax:
                currentWinner = language
                currentMax = count
        
        return currentWinner
    


pytagcloud/__init__.py
# -*- coding: utf-8 -*-
from copy import copy
from math import sin, cos
from pygame import transform, font, mask, Surface, Rect, SRCALPHA, draw
from pygame.sprite import Group, Sprite, collide_mask
from random import randint, choice
import colorsys
import os
import pygame
import json
import logging


TAG_PADDING = 5
STEP_SIZE = 2  # relative to base step size of each spiral function
RADIUS = 1
ECCENTRICITY = 1.5

LOWER_START = 0.45
UPPER_START = 0.55

FONT_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'fonts')
DEFAULT_FONT = 'Droid Sans'
DEFAULT_PALETTE = 'default'
fd = open(os.path.join(FONT_DIR, 'fonts.json'), 'r')
FONT_CACHE = json.loads(fd.read())

# pygame.init()

convsurf = Surface((2 * TAG_PADDING, 2 * TAG_PADDING))
convsurf.fill((255, 0, 255))
convsurf.set_colorkey((255, 0, 255))
draw.circle(convsurf, (0, 0, 0), (TAG_PADDING, TAG_PADDING), TAG_PADDING)
CONVMASK = mask.from_surface(convsurf)

LAYOUT_HORIZONTAL = 0
LAYOUT_VERTICAL = 1
LAYOUT_MOST_HORIZONTAL = 2
LAYOUT_MOST_VERTICAL = 3
LAYOUT_MIX = 4
LAYOUT_FORTYFIVE = 5
LAYOUT_RANDOM = 6

LAYOUTS = (
    LAYOUT_HORIZONTAL,
    LAYOUT_VERTICAL,
    LAYOUT_MOST_HORIZONTAL,
    LAYOUT_MOST_VERTICAL,
    LAYOUT_MIX,
    LAYOUT_FORTYFIVE,
    LAYOUT_RANDOM
)

LAST_COLLISON_HIT = None


'''
cloud_canvas = None


def get_cloud_cr(canvas=None):
    global cloud_canvas

    if cloud_canvas is None:
        cr = canvas.get_property('window').cairo_create()
        canvas.connect('draw', self.__draw_cb)

        cloud_canvas = cr.get_target().create_similar(
            cairo.CONTENT_COLOR, size, size)

    _cr = cairo.Context(cloud_canvas)
    return _cr
'''


def __draw_cb(self, canvas, cr):
    cr.set_source_surface(cloud_canvas)
    cr.paint()


class Tag(Sprite):
    """
    Font tag sprite. Blit the font to a surface to correct the font padding
    """

    def __init__(self, tag, initial_position, fontname=DEFAULT_FONT):
        Sprite.__init__(self)
        self.tag = copy(tag)
        self.rotation = 0

        self.font_spec = load_font(fontname)
        self.font = font.Font(os.path.join(FONT_DIR,
                                           self.font_spec['ttf']),
                              self.tag['size'])
        # fonter = self.font.render(unicode(tag['tag'], 'UTF-8'), True,
        fonter = self.font.render(tag['tag'], True, tag['color'])
        frect = fonter.get_bounding_rect()
        frect.x = -frect.x
        frect.y = -frect.y
        self.fontoffset = (-frect.x, -frect.y)
        font_sf = Surface((frect.width, frect.height), pygame.SRCALPHA, 32)
        font_sf.blit(fonter, frect)
        self.image = font_sf
        self.rect = font_sf.get_rect()
        self.rect.width += TAG_PADDING
        self.rect.height += TAG_PADDING
        self.rect.x = initial_position[0]
        self.rect.y = initial_position[1]
        self._update_mask()

    def _update_mask(self):
        self.mask = mask.from_surface(self.image)
        self.mask = self.mask.convolve(CONVMASK, None,
                                       (TAG_PADDING, TAG_PADDING))

    def flip(self):
        angle = 90 if self.rotation == 0 else - 90
        self.rotate(angle)

    def rotate(self, angle):
        pos = (self.rect.x, self.rect.y)
        self.image = transform.rotate(self.image, angle)
        self.rect = self.image.get_rect()
        self.rect.x, self.rect.y = pos
        self._update_mask()

    def update_fontsize(self):
        self.font = font.Font(os.path.join(FONT_DIR, self.font_spec['ttf']),
                              self.tag['size'])


def load_font(name):
    for font in FONT_CACHE:
        if font['name'].encode('utf-8') == name.encode('utf-8'):
            return font
    raise AttributeError('Invalid font name. Should be one of %s' %
                         ", ".join([f['name'] for f in FONT_CACHE]))


def defscale(count, mincount, maxcount, minsize, maxsize):
    if maxcount == mincount:
        return int((maxsize - minsize) / 2.0 + minsize)
    return int(minsize + (maxsize - minsize) *
               (count * 1.0 / (maxcount - mincount)) ** 0.8)


def make_tags(wordcounts, minsize=3, maxsize=36, colors=None, scalef=defscale):
    """
    sizes and colors tags
    wordcounts is a list of tuples(tags, count). (e.g. how often the
    word appears in a text)
    the tags are assigned sizes between minsize and maxsize, the function used
    is determined by scalef (default: square root)
    color is either chosen from colors (list of rgb tuples) if provided or
    random
    """
    counts = [tag[1] for tag in wordcounts]

    if not len(counts):
        return []

    maxcount = max(counts)
    mincount = min(counts)
    tags = []
    for word_count in wordcounts:
        color = choice(colors) if colors else (randint(10, 220),
                                               randint(10, 220),
                                               randint(10, 220))
        tags.append({'color': color,
                     'size': scalef(word_count[1], mincount,
                                    maxcount, minsize, maxsize),
                     'tag': word_count[0]})
    logging.debug('tag count={}'.format(len(tags)))
    logging.debug(tags)
    return tags


def _do_collide(sprite, group):
    """
    Use mask based collision detection
    """
    global LAST_COLLISON_HIT
    # Test if we still collide with the last hit
    if LAST_COLLISON_HIT and collide_mask(sprite, LAST_COLLISON_HIT):
        return True

    for sp in group:
        if collide_mask(sprite, sp):
            LAST_COLLISON_HIT = sp
            return True
    return False


def _get_tags_bounding(tag_store):
    if not len(tag_store):
        return Rect(0, 0, 0, 0)
    rects = [tag.rect for tag in tag_store]
    return rects[0].unionall(rects[1:])


def _get_group_bounding(tag_store, sizeRect):
    if not isinstance(sizeRect, pygame.Rect):
        sizeRect = Rect(0, 0, sizeRect[0], sizeRect[1])
    if tag_store:
        rects = [tag.rect for tag in tag_store]
        union = rects[0].unionall(rects[1:])
        if sizeRect.contains(union):
            return union
    return sizeRect


def _archimedean_spiral(reverse):
    DEFAULT_STEP = 0.05  # radians
    t = 0
    r = 1
    if reverse:
        r = -1
    while True:
        t += DEFAULT_STEP * STEP_SIZE * r
        yield (ECCENTRICITY * RADIUS * t * cos(t), RADIUS * t * sin(t))


def _rectangular_spiral(reverse):
    DEFAULT_STEP = 3  # px
    directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
    if reverse:
        directions.reverse()
    direction = directions[0]

    spl = 1
    dx = dy = 0
    while True:
        for step in range(spl * 2):
            if step == spl:
                direction = directions[(spl - 1) % 4]
            dx += direction[0] * STEP_SIZE * DEFAULT_STEP
            dy += direction[1] * STEP_SIZE * DEFAULT_STEP
            yield dx, dy
        spl += 1


def _search_place(current_tag, tag_store, canvas, spiral, ratio):
    """
    Start a spiral search with random direction.
    Resize the canvas if the spiral exceeds the bounding rectangle
    """

    reverse = choice((0, 1))
    start_x = current_tag.rect.x
    start_y = current_tag.rect.y
    min_dist = None
    opt_x = opt_y = 0

    current_bounding = _get_tags_bounding(tag_store)
    cx = current_bounding.w / 2.0
    cy = current_bounding.h / 2.0

    for dx, dy in spiral(reverse):
        current_tag.rect.x = start_x + dx
        current_tag.rect.y = start_y + dy
        if not _do_collide(current_tag, tag_store):
            if canvas.contains(current_tag.rect):
                tag_store.add(current_tag)
                return
            else:
                # get the distance from center
                current_dist = (abs(cx - current_tag.rect.x) ** 2 +
                                abs(cy - current_tag.rect.y) ** 2) ** 0.5
                if not min_dist or current_dist < min_dist:
                    opt_x = current_tag.rect.x
                    opt_y = current_tag.rect.y
                    min_dist = current_dist

                # only add tag if the spiral covered the canvas boundaries
                if abs(dx) > canvas.width / 2.0 and \
                   abs(dy) > canvas.height / 2.0:
                    current_tag.rect.x = opt_x
                    current_tag.rect.y = opt_y
                    tag_store.add(current_tag)

                    new_bounding = current_bounding.union(current_tag.rect)

                    delta_x = delta_y = 0.0
                    if new_bounding.w > canvas.width:
                        delta_x = new_bounding.w - canvas.width

                        canvas.width = new_bounding.w
                        delta_y = ratio * new_bounding.w - canvas.height
                        canvas.height = ratio * new_bounding.w

                    if new_bounding.h > canvas.height:
                        delta_y = new_bounding.h - canvas.height

                        canvas.height = new_bounding.h
                        canvas.width = new_bounding.h / ratio
                        delta_x = canvas.width - canvas.width

                    # realign
                    for tag in tag_store:
                        tag.rect.x += delta_x / 2.0
                        tag.rect.y += delta_y / 2.0

                    canvas = _get_tags_bounding(tag_store)
                    return


def _draw_cloud(tag_list,
                layout=LAYOUT_MIX,
                size=(500, 500),
                fontname=DEFAULT_FONT,
                rectangular=False):

    # sort the tags by size and word length
    tag_list.sort(key=lambda tag: len(tag['tag']))
    tag_list.sort(key=lambda tag: tag['size'])
    tag_list.reverse()

    # create the tag space
    tag_sprites = []
    area = 0
    for tag in tag_list:
        tag_sprite = Tag(tag, (0, 0), fontname=fontname)
        area += tag_sprite.mask.count()
        tag_sprites.append(tag_sprite)

    canvas = Rect(0, 0, 0, 0)
    ratio = float(size[1]) / size[0]

    if rectangular:
        spiral = _rectangular_spiral
    else:
        spiral = _archimedean_spiral

    aligned_tags = Group()
    for tag_sprite in tag_sprites:
        angle = 0
        if layout == LAYOUT_MIX and randint(0, 1) == 0:
            angle = 90
        elif layout == LAYOUT_VERTICAL:
            angle = 90
        elif layout == LAYOUT_FORTYFIVE:
            if randint(0, 1) == 0:
                angle = 45
            else:
                angle = 315
        elif layout == LAYOUT_RANDOM:
            angle = randint(0, 89)

        tag_sprite.rotate(angle)

        xpos = canvas.width - tag_sprite.rect.width
        if xpos < 0:
            xpos = 0
        xpos = randint(int(xpos * LOWER_START), int(xpos * UPPER_START))
        tag_sprite.rect.x = xpos

        ypos = canvas.height - tag_sprite.rect.height
        if ypos < 0:
            ypos = 0
        ypos = randint(int(ypos * LOWER_START), int(ypos * UPPER_START))
        tag_sprite.rect.y = ypos

        _search_place(tag_sprite, aligned_tags, canvas, spiral, ratio)

    canvas = _get_tags_bounding(aligned_tags)

    # resize cloud
    zoom = min(float(size[0]) / canvas.w, float(size[1]) / canvas.h)

    logging.debug('zoom {}'.format(zoom))
    cw = int(canvas.w * zoom)
    ch = int(canvas.h * zoom)

    for tag in aligned_tags:
        tag.rect.x *= zoom
        tag.rect.y *= zoom
        tag.rect.width *= zoom
        tag.rect.height *= zoom

        if tag.rect.x + tag.rect.width > cw:
            tag.rect.x = cw - tag.rect.width
        if tag.rect.y + tag.rect.height > ch:
            tag.rect.y = ch - tag.rect.height

        tag.tag['size'] = int(tag.tag['size'] * zoom)
        tag.update_fontsize()

    canvas = _get_tags_bounding(aligned_tags)

    padding = 160
    canvas.width += padding
    canvas.height += padding

    return canvas, aligned_tags


def create_tag_image(
        tags,
        output,
        size=(1024, 768),
        background=(255, 255, 255),
        layout=LAYOUT_MIX,
        fontname=DEFAULT_FONT,
        rectangular=False):
    """
    Create a png tag cloud image
    """

    if not len(tags):
        return

    pygame.init()

    sizeRect, tag_store = _draw_cloud(tags,
                                      layout,
                                      size=size,
                                      fontname=fontname,
                                      rectangular=rectangular)

    tag_surface = Surface((sizeRect.w, sizeRect.h), SRCALPHA, 32)
    tag_surface.fill(background)
    for tag in tag_store:
        tag_surface.blit(tag.image, tag.rect)
    tag_surface = pygame.transform.scale(tag_surface,
                                         (int(sizeRect.w * 0.9),
                                          int(sizeRect.h * 0.9)))

    xo = int(sizeRect.w * 0.05)
    yo = int(sizeRect.h * 0.05)
    output_surface = Surface((sizeRect.w, sizeRect.h), SRCALPHA, 32)
    output_surface.fill(background)
    output_surface.blit(tag_surface, (xo, yo))
    pygame.image.save(output_surface, output)

    pygame.quit()


def create_html_data(tags,
                     size=(1024, 768),
                     layout=LAYOUT_MIX,
                     fontname=DEFAULT_FONT,
                     rectangular=False):
    """
    Create data structures to be used for HTML tag clouds.
    """

    if not len(tags):
        return

    sizeRect, tag_store = _draw_cloud(tags,
                                      layout,
                                      size=size,
                                      fontname=fontname,
                                      rectangular=rectangular)

    tag_store = sorted(tag_store, key=lambda tag: tag.tag['size'])
    tag_store.reverse()
    data = {
        'css': {},
        'links': []
    }

    color_map = {}
    for color_index, tag in enumerate(tags):
        if not tag['color'] in color_map:
            color_name = 'c{}'.format(color_index)
            hslcolor = colorsys.rgb_to_hls(tag['color'][0] / 255.0,
                                           tag['color'][1] / 255.0,
                                           tag['color'][2] / 255.0)
            lighter = hslcolor[1] * 1.4
            if lighter > 1:
                lighter = 1
            light = colorsys.hls_to_rgb(hslcolor[0], lighter, hslcolor[2])
            data['css'][color_name] = ('#%02x%02x%02x' % tag['color'],
                                       '#%02x%02x%02x' % (light[0] * 255,
                                                          light[1] * 255,
                                                          light[2] * 255))
            color_map[tag['color']] = color_name

    for stag in tag_store:
        line_offset = 0

        line_offset = stag.font.get_linesize() - \
            (stag.font.get_ascent() + abs(stag.font.get_descent()) -
             stag.rect.height) - 4

        tag = {
            'tag': stag.tag['tag'],
            'cls': color_map[stag.tag['color']],
            'top': stag.rect.y - sizeRect.y,
            'left': stag.rect.x - sizeRect.x,
            'size': int(stag.tag['size'] * 0.85),
            'height': int(stag.rect.height * 1.19) + 4,
            'width': stag.rect.width,
            'lh': line_offset
        }

        data['links'].append(tag)
        data['size'] = (sizeRect.w, sizeRect.h * 1.15)

    return data
